### ISSUE TITLE
Inconsinstent search behavior with german language and umlauts

### ISSUE THREAD

**Issue by Jaykju** [squidfunk/mkdocs-material#1603]
__I checked that...__

- [x] ... the documentation does not mention anything about my problem
- [x] ... the problem doesn't occur with the default MkDocs template
- [x] ... the problem is not in any of my customizations (CSS, JS, template)
- [x] ... there are no open or closed issues that are related to my problem

## Description

The search does not display results when searching for special characters. In this specific case I am interested in searching for ä, ö and ü in a german documentation.
The stemming (if I understood it correctly) also seems to prevent the right search results from appearing.

~Our documentation is currently available at [herbs-lit.jaykju.de](https://herbs-lit.jaykju.de) if you want to check on the behavior~ (I disabled the transform function for now, because it seems to yield better results). The [repository](https://github.com/Bettlaken/K_H_Literature) is linked there as well.

### Expected behavior

Display search results as usual. If the exact word is in the text it should appear as a search result. 

### Actual behavior

When starting to search a word the results appear irregularly.

Some examples in our documentation for a herbs iot project:

Searching for "Kr" yields the following result:
![image](https://user-images.githubusercontent.com/33766870/79281071-5a26eb80-7eb2-11ea-8ecd-46801caa1d03.png)

Adding the Character "ä" shows as follows:
![image](https://user-images.githubusercontent.com/33766870/79281117-6f037f00-7eb2-11ea-988f-311bcf68fee0.png)

Finishing the word still does not yield any results:
![image](https://user-images.githubusercontent.com/33766870/79281530-36b07080-7eb3-11ea-804c-a148c4cdb94f.png)

Adding a second search term does however yield results again:
![image](https://user-images.githubusercontent.com/33766870/79281570-5182e500-7eb3-11ea-87b1-23edc7830cd3.png)

These results however only last until the stemmer kicks in and in this case removes the "en" from anpflanzen in the index, which is probably the reason why this is not working:
![image](https://user-images.githubusercontent.com/33766870/79281668-8abb5500-7eb3-11ea-9c84-3dc7b14603a7.png)

### Steps to reproduce the bug

Search for strings containing special characters like ä, ö and ü in a documentation that contains them.

I also tried overriding the transform function and found that the default regular expression does not append a * to "krä" but even changing it to support ä, ö and ü did not help. 
The regular expression I tried was the following:

```html
<script>
    var search = {
      transform: function(query) {
        value = query.replace(/(?:^|\s+)[*+-:^~]+(?=\s+|$)/g, "") 
            .trim()
            .replace(/\s+|\b([äöü]*)$/g, "$1* ");
          console.log(value);
          return value;
        }
    }
</script>
```

But it seems to yield even worse results. Omitting the second replace seems to work for different cases but not for others as well. 

### Package versions

* Python: 3.8
* MkDocs: 1.1
* Material: 5.1.0

### Project configuration

``` yaml
# Configuration
theme:
  name: material
  language: de

# Plugins
plugins:
  - search:
      lang: de
  - tags:
      tags_folder: tags
      tags_template: theme/tags.md.template

# Extensions
markdown_extensions:
  - meta
  - admonition
  - codehilite:
      guess_lang: false
  - toc:
      permalink: true
```

Or look [here](https://github.com/Bettlaken/K_H_Literature/blob/master/mkdocs.yml)

### System information

* OS: Windows 10 | Ubuntu 16.04
* Browser: Chrome

I would appreciate any insights or suggestions.

**Comment by squidfunk** [comment]
I’m pretty sure that the problem lies in the word boundary in the second replacement, resulting in `Kr*ä*`. Does it work if you remove `\b` from the Regex?

**Comment by squidfunk** [comment]
As a further note, if the problem doesn't occur with the default template, you can just override the transform to return the input value as-is to get the behavior of the default theme.

**Comment by Jaykju** [comment]
Using my edited transform function the following queries with their respective amount of results are created:

![image](https://user-images.githubusercontent.com/33766870/79306887-49986480-7ef6-11ea-8d7e-f16c83ef6002.png)

I think this matches the expected behaviour. The default transform function of Material works the same except for the third entry, where it does not append the * to Krä. 

> I’m pretty sure that the problem lies in the word boundary in the second replacement, resulting in Kr*ä*. Does it work if you remove \b from the Regex?

I tried but afterwards no asterisks where added to the query anymore.

> As a further note, if the problem doesn't occur with the default template, you can just override the transform to return the input value as-is to get the behavior of the default theme.

This is the current solution but that means that I only get exact matches for words and not the nice search functionality that Material otherwise provides.

**Comment by squidfunk** [comment]
I see, so removing (or replacing) the word boundary seems to fix at least part of the issue. That's something we can address. Regarding stemming – you can disable stemming by extending the theme and overriding `partials/language/de.html`, specifically `search.config.pipeline` to an empty string:

https://github.com/squidfunk/mkdocs-material/blob/572c1e854ef87dd8e985d7c6a62a30d832422da9/src/partials/language/en.html#L38

**Comment by pschoenmakers** [comment]
Don't want to hijack this thread, but as my problem seems totally related didn't want to open an new issue either.

This is with Dutch as language, result for "opzegg"
![Bildschirmfoto 2020-04-15 um 10 26 38](https://user-images.githubusercontent.com/15877946/79315781-bc5c0c80-7f03-11ea-98f8-44b88152517c.jpg)

Then adding "en" to the word does not return those exact matches at all.
![Bildschirmfoto 2020-04-15 um 10 26 44](https://user-images.githubusercontent.com/15877946/79316150-3ab8ae80-7f04-11ea-872d-fff4f56b966e.jpg)

did add this to mkdocs-material/src/partials/language/nl.html
`"search.config.pipeline": "",`

**Comment by squidfunk** [comment]
Search is super difficult to get right across 40 languages and so many different use cases. I guess we could try to improve the current behavior, but I wouldn't consider it a super high priority issue. Again, if someone wants to go down that path, I'm happy to assist, but I have to say that this can be quite a rabbithole.

**Comment by Jaykju** [comment]
I overwrote the de.html partial and set `"search.config.pipeline": "",` as well, but the result did not change either. 

> if someone wants to go down that path, I'm happy to assist, but I have to say that this can be quite a rabbit hole.

I can definitely see that. I think I am going to have to do a little more research to figure out why exactly there is a problem with umlauts first because I am not sure if it might be lunr related.

**Comment by squidfunk** [comment]
> I think I am going to have to do a little more research to figure out why exactly there is a problem with umlauts first because I am not sure if it might be lunr related.

That is definitely a good idea. The aim should be to isolate whether there's a problem with lunr or the theme. We can adjust the transform, but we should better be sure that we don't introduce problems with other languages. The default mkdocs theme doesn't pre-process the query, so it's also worth trying out in order to better understand the behavior.

**Comment by Jaykju** [comment]
@pschoenmakers 

Overwriting the config block in the main.html to remove the transform function did fix that behaviour for me.

main.html contents:
```html
{% extends "base.html" %}

{% block config %}
  <script>
    var search = {
      transform: function(query) {
        return query;
      }
    }
  </script>
{% endblock %}
```

For me this yields exact matches +-2chars.
![image](https://user-images.githubusercontent.com/33766870/79363921-d835d180-7f48-11ea-8146-50a536888ccd.png)

**Comment by squidfunk** [comment]
As I said, that is the exact behavior of the default MkDocs theme. Now we have to find a way to preserve partial matches while respecting stemming.

**Comment by Jaykju** [comment]
Im not sure if this is an overly hacky way but what about duplicating the word and appending an asterisk to one of them? 

Such that the query `anpfla` would be converted to `anpfla* anpfla`. The first term yielding partial matches and the second yielding exact matches. 

This also works with the full word `anpflanzen` because the second part of the query `anpflanzen* anpflanzen` would match the exact word.

**Comment by squidfunk** [comment]
That'll probably work for your use case but might not be a general solution.

**Comment by pschoenmakers** [comment]
> Overwriting the config block in the main.html to remove the transform function did fix that behaviour for me.

This does indeed help. The problem is really somewhere in the implementation of the material theme as other themes do work fine.

**Comment by squidfunk** [comment]
It works fine in the stemming case, but you'll lose partial matching. My guess is, that this is not theme-specific but more of a general lunr.js problem. Material for MkDocs tries to provide a good typeahead experience, which may clash with stemming. As noted, you may override the transform function if you wish to give up on the typeahead in favor of stemming. You may also just disable stemming.

Maybe there's a solution which caters to both cases, but, as said, it's a rabbit hole for which I'm lacking the time to investigate. Nevertheless, it's super easy to override and revert back to the default behavior.

**Comment by squidfunk** [comment]
I decided to close this issue for two reasons.

1. When encountering this theme-specific issue, the default behavior of MkDocs can be revert to by simply definition the `transform` function as the original OP already suggested.

2. It's unlikely that we will find a configuration for search which caters to all cases. Either we lose stemming support or typeahead. To my best knowledge, there's no way to get the best of both. If somebody wants to go down this rabbit hole and manages to improve the search, I'd be super happy to merge it. Until then, this is the default behavior, that can be completely customized through the bespoke `transform` function.

**Comment by squidfunk** [comment]
As noted in https://github.com/squidfunk/mkdocs-material/issues/1745#issuecomment-640206622, search should now yield better results/behavior for languages other than English as of 36e1e3d. The change is released as part of 5.2.3.
